### DISCUSSION TITLE
Project Structure: How can App.swift import ComposableArchitecture?

### DISCUSSION THREAD

**Discussion by oco-adam** [pointfreeco/isowords#141] [Q&A] [answered]
Hi,

I'm new to Swift and trying to understand the project structure. 

**In particular, I don't understand how the file /iOS/App.swift can import the module ComposableArchitecture?** 

It doesn’t seem to be specified as a dependency anywhere for the isowords target - when I clone the repo I see the following as dependencies for the isowords target:

![Screenshot 2021-12-12 at 18 42 37](https://user-images.githubusercontent.com/32979919/145724408-29243485-4236-4704-bb2e-711ac27e1805.png)

I've tried to set up a minimal project (in Xcode 13.1) using the same basic structure of isowords, but when I import and use ComposableArchitecture in my /iOS/App.swift file, I get errors of the form:
Undefined symbol: nominal type descriptor for ComposableArchitecture.Store

This might quite probably be due to a basic misunderstanding on my part of Xcode, SPM, Swift, but any help would be much appreciated.

**Comment by mbrandonw** [comment, marked as answer]
Hi @oco-adam, the reason we can import ComposableArchitecture in App.swift (as well as all of [these imports](https://github.com/pointfreeco/isowords/blob/b87f4c61c0523564adf17c031f9fa4b88a40d631/App/iOS/App.swift#L1-L13)) is because Xcode/SPM picks up transitive dependencies. So, since AppFeature depends on ComposableArchitecture, and the app target depends on AppFeature, we get to import all TCA in the app target.

**Comment by gohanlon** [comment]
Hi @oco-adam, @mbrandonw beat me to it. I'll add that while isoword's dependencies are specified using SPM, they aren't specified directly inside Xcode, which is how most people are familiar with using SPM. Instead, isowords uses a (quite large) [`Package.swift`](https://github.com/pointfreeco/isowords/blob/main/Package.swift) file. And when Xcode sees a `Package.swift` file, Xcode's build system will "magically" use it. This is unintuitive yet delightful.

I have adopted the "isowords structure" in my own projects, and it sounds like you may be undertaking that task as well. I found Point-Free's video series covering isowords to be extremely useful in explaining the hows and whys of the isowords project structure, as well as many other aspects of the project:

* [A Tour of isowords: Part 1](https://www.pointfree.co/episodes/ep142-a-tour-of-isowords-part-1) ← Covers the SPM-related project structure
* [A Tour of isowords: Part 2](https://www.pointfree.co/episodes/ep143-a-tour-of-isowords-part-2)
* [A Tour of isowords: Part 3](https://www.pointfree.co/episodes/ep144-a-tour-of-isowords-part-3)
* [A Tour of isowords: Part 4](https://www.pointfree.co/episodes/ep145-a-tour-of-isowords-part-4)

Cheers!

**Comment by oco-adam** [comment]
Hi @mbrandonw and @gohanlon thanks very much for your quick replies!

I wondered if this might be the case with transitive dependencies, so thanks for confirming this. 

However this confirmed to me that my App.swift should have already been using this dependency transitively.

So following your answer, I just started to give details of my project because my App.swift was not picking up the transitive dependency from one of the modules from my "hyper-modularisation" package.

However when started copying the Package.swift to describe here the details of my problem, I realised that maybe I needed to specify the dependency in two places, one in the dependencies property and once in the target property for MyApp?:

So when I added the line
```
.product(name: "ComposableArchitecture", package: "swift-composable-architecture")
```

to my Package.swift (below), the transitive dependency got picked up and everything is working now.


```
/// MyApp/Package.swift (which will contain all the small modules)
// swift-tools-version:5.5
// The swift-tools-version declares the minimum version of Swift required to build this package.

import PackageDescription

let package = Package(
    name: "MyApp",
    platforms: [
      .iOS(.v13)
    ],
    products: [
        // Products define the executables and libraries a package produces, and make them visible to other packages.
        .library(
            name: "MyApp",
            targets: ["MyApp"]),
    ],
    dependencies: [
        // Dependencies declare other packages that this package depends on.
        // .package(url: /* package url */, from: "1.0.0"),
        .package(url: "https://github.com/pointfreeco/swift-composable-architecture", from: "0.29.0")
    ],
    targets: [
        // Targets are the basic building blocks of a package. A target can define a module or a test suite.
        // Targets can depend on other targets in this package, and on products in packages this package depends on.
        .target(
            name: "MyApp",
            dependencies: [
	.product(name: "ComposableArchitecture", package: "swift-composable-architecture"),
]),
        .testTarget(
            name: "MyAppTests",
            dependencies: ["MyApp"]),
    ]
)
```

So maybe my issue was caused by the need to always specify the dependencies in the Package.swift in two places.

I did actually watch the videos @gohanlon refers to and maybe this was stated somewhere there, but I thought I'd post this here in case it helps out another Swift newcomer.

Anyway, thanks for your help!
